### PR TITLE
Initial implementation of context menu for save as.

### LINKED ISSUE #11: 3: Convert dictionaries: right-click dictionary and select "save as..." with RTF and JSON options.

### PR DESCRIPTION
**Pull request by sofa13**
Fixes #11

The fix is not complete yet, I'm putting in a pull request now as a way to show progress.
I've only worked on getting the menu to show and haven't started implementing the functionality. 
Let me know if you see any red flags along the way.

### CODE REVIEW

**Comment by sofa13** [comment]
Issues I'm looking into:
- on cancel from file dialogue, the whole plover application closes instead of just the file dialogue.
-  the conversion is working, but upon automatically adding the converted dictionary and updating the dictionary list, the converted dictionary appears blank when edited. This does not occur if the new dictionary is the same file type.

**Comment by sofa13** [comment]
That's my completed implementation, all the issues that I noticed have been resolved. 
Let me know what you guys think, then I'll look into writing test cases.

**Comment by panathea** [comment]
- The context menu shows up at a weird location when I right-click on a row
- Save As text usually has an ellipsis, either `...` or `…`
- The dictionary file selector is too loose. It would be better to tighten it up a little, maybe try this:

``` python
        filters = ';;'.join('%s Dictionary (*%s)' % (ext.upper().strip('.'), ext) for ext in dictionary_formats)
        new_filename = QFileDialog.getSaveFileName(
            self, _('Save Dictionary'), filename,
            filters,
        )[0]
```

Note: the default type will need to be set to the dictionary's current type.
- The dialog box's message after conversion is too vague.

`custom.rtf was saved to ~/git/dictionary. Would you like to add the new dictionary to Plover?`
- The dialog's title should have a meaningful name like `Dictionary Saved`
- Not sure if it makes a difference, but `QMessageBox.Yes|QMessageBox.No` instead of multiple arguments might take away the implication of order. It would be nice if the default action were `Yes` as well.
- Another action for the context menu might just be an `Open containing folder`, it would be simple if you want to add it to this PR (or not)

**Comment by sofa13** [comment]
All feedback above has been implemented except for the first one (need a bit more time to look into this one), and the last one (need a bit more time to implement this, but I agree I'll include it in this PR).

File: plover/gui_qt/dictionaries_widget.py [hunk 1]
```diff
@@ -256,3 +263,56 @@ def on_move_dictionaries_down(self):
...
+            basename = os.path.basename(filename)
+            ext_new = os.path.splitext(new_filename)[-1].lower()
+
+            if file_ext == ext_new:
+                shutil.copyfile(filename, new_filename)
+            else:
+                open(new_filename, "w")
+                self.convert_dictionary(filename, new_filename)
+
+            msg = basename + " was saved to " + new_filename + ". Would you like to add the new dictionary to Plover?"
```
Comment (panathea) [inline plover/gui_qt/dictionaries_widget.py]: -            msg = basename + " was saved to " + new_filename + ". Would you like to add the new dictionary to Plover?" 

`msg = '%s was saved to %s\n\nWould you like to add the new dictionary to Plover?' % (file_name, file_directory)`

Mm, also code style dictates that we use `''` for strings, not `""`

**Comment by sofa13** [comment]
- dialog's message about conversion has been fixed
- context menu position has been improved
- open file directory action has been implemented

**Comment by panathea** [comment]
This patch:
- Fixes open containing folder on Mac
- Uses the phrase "Open Containing Folder"
- Uses `' '` for strings
- Moves convert dictionary into dictionary base
- Handles errors if the conversion excepts

``` diff
diff --git a/plover/dictionary/base.py b/plover/dictionary/base.py
index 4593745..994ac71 100644
--- a/plover/dictionary/base.py
+++ b/plover/dictionary/base.py
@@ -78,7 +78,15 @@ def save_dictionary(d, filename, saver):

     # Then move the new file to the final location.
     shutil.move(tmp, filename)
-    
+
+
+def convert_dictionary(read_path, write_path):
+    dict_in = load_dictionary(read_path)
+    dict_out = create_dictionary(write_path)
+    dict_out.update(dict_in)
+    dict_out.save()
+
+
 class ThreadedSaver(object):
     """A callable that saves a dictionary in the background.

diff --git a/plover/gui_qt/dictionaries_widget.py b/plover/gui_qt/dictionaries_widget.py
index 85d7e0f..7a94729 100644
--- a/plover/gui_qt/dictionaries_widget.py
+++ b/plover/gui_qt/dictionaries_widget.py
@@ -3,6 +3,7 @@ import os
 import shutil
 import webbrowser

+import sys
 from PyQt5.QtCore import (
     QItemSelection,
     QItemSelectionModel,
@@ -19,13 +20,14 @@ from PyQt5.QtWidgets import (
     QMessageBox
 )

+from plover import log
 from plover.dictionary.base import dictionaries as dictionary_formats
 from plover.misc import shorten_path

 from plover.gui_qt.dictionaries_widget_ui import Ui_DictionariesWidget
 from plover.gui_qt.dictionary_editor import DictionaryEditor
 from plover.gui_qt.utils import ToolBar
-from plover.dictionary.base import create_dictionary, load_dictionary
+from plover.dictionary.base import convert_dictionary


 class DictionariesWidget(QWidget, Ui_DictionariesWidget):
@@ -266,63 +268,71 @@ class DictionariesWidget(QWidget, Ui_DictionariesWidget):
         self._set_selection(selection)

     def _row_selector(self, pos):
-        vheader = self.table.verticalHeader()
-        vheader.setContextMenuPolicy(Qt.CustomContextMenu)
-        position = vheader.mapToGlobal(pos)
         menu = QMenu()

-        row_action = QAction("Save As...", self)
-        row_action_open = QAction("Open...", self)
+        action = QAction('Save As...', self)
+        action.triggered.connect(self._save_as)
+        menu.addAction(action)

-        row_action.triggered.connect(self.save_as)
-        row_action_open.triggered.connect(self.open_file_directory)
+        action = QAction('Open Containing Folder', self)
+        action.triggered.connect(self._open_file_directory)
+        menu.addAction(action)

-        menu.addAction(row_action)
-        menu.addAction(row_action_open)
+        menu.exec(self.table.viewport().mapToGlobal(pos))

-        menu.exec(position)
-
-    def open_file_directory(self):
+    def _open_file_directory(self):
         file_name = [self._dictionaries[item.row()]
                     for item in self.table.selectedItems()][0]
         file_directory = os.path.dirname(file_name)
+        if sys.platform.startswith('darwin'):
+            file_directory = 'file://%s' % file_directory
         webbrowser.open(file_directory)

-    def save_as(self):
+    def _save_as(self):
         filename = [self._dictionaries[item.row()]
                         for item in self.table.selectedItems()][0]
         file_ext = os.path.splitext(filename)[-1].lower()
         if file_ext in dictionary_formats:
-            selectedFilter = '%s Dictionary (*%s)' % (file_ext.upper().strip('.'), file_ext)
+            selected_filter = '%s Dictionary (*%s)' % (
+                file_ext.upper().strip('.'), file_ext
+            )
+
+        filters = ';;'.join('%s Dictionary (*%s)' % (
+            ext.upper().strip('.'), ext) for ext in dictionary_formats)

-        filters = ';;'.join('%s Dictionary (*%s)' % (ext.upper().strip('.'), ext) for ext in dictionary_formats)
         new_filename = QFileDialog.getSaveFileName(
-            self, _('Save Dictionary'), filename,
-            filters, selectedFilter
+            self, _('Save Dictionary As...'), filename,
+            filters, selected_filter
         )[0]

         if new_filename:
             file_name = os.path.basename(new_filename)
             file_directory = os.path.dirname(new_filename)
-            ext_new = os.path.splitext(new_filename)[-1].lower()
+            new_ext = os.path.splitext(new_filename)[-1].lower()

-            if file_ext == ext_new:
-                shutil.copyfile(filename, new_filename)
+            try:
+                if file_ext == new_ext:
+                    shutil.copyfile(filename, new_filename)
+                else:
+                    open(new_filename, 'w')
+                    convert_dictionary(filename, new_filename)
+            except Exception:
+                log.error(
+                    'Error while saving new dictionary: %s',
+                    new_filename,
+                    exc_info=True
+                )
             else:
-                open(new_filename, "w")
-                self.convert_dictionary(filename, new_filename)
-
-            msg = '%s was saved to %s. Would you like to add the new dictionary to Plover?' % (file_name, file_directory)
-            reply = QMessageBox.question(self, 'Dictionary Saved',
-                msg, QMessageBox.Yes | QMessageBox.No)
-            if reply == QMessageBox.Yes:
-                dictionaries_list = list(self._dictionaries)
-                if new_filename not in dictionaries_list:
-                    dictionaries_list.append(new_filename)
-                self._update_dictionaries(dictionaries_list)
-
-    def convert_dictionary(self, input, output):
-        id = load_dictionary(input)
-        od = create_dictionary(output)
-        od.update(id)
-        od.save()
\ No newline at end of file
+                msg = ('%s was saved to %s.\n\n'
+                       'Would you like to add the new dictionary to Plover?'
+                        % (file_name, file_directory)
+                       )
+                if QMessageBox.question(
+                    self, 'Dictionary Saved', msg,
+                    QMessageBox.Yes | QMessageBox.No,
+                    QMessageBox.Yes
+                ) == QMessageBox.Yes:
+                    dictionaries = list(self._dictionaries)
+                    if new_filename not in self._dictionaries:
+                        dictionaries.append(new_filename)
+                    self._update_dictionaries(dictionaries)
```

**Comment by panathea** [comment]
Thinking about it, eventually we're going to have to add some more context menu entries for the existing actions:
- Remove dictionary
- Edit dictionary…

And maybe reordering? I don't know if reordering is worth the effort, but definitely some people will tend to gravitate to the context menu to see what's what.

This is obviously not a requirement for this pull request but we should be aware of it.

**Pull request merged**